src/ra_log_reader.erl

```diff
@@ -170,13 +174,55 @@ sparse_read(#?STATE{cfg = #cfg{} = Cfg} = State, Indexes, Entries0) ->
     ok = incr_counter(Cfg, ?C_RA_LOG_READ_SEGMENT, SegC),
     {Entries, State#?MODULE{open_segments = Open}}.
 
+-spec read_plan(state(), [ra_index()]) -> read_plan().
+read_plan(#?STATE{segment_refs = SegRefs}, Indexes) ->
+    %% TODO: add counter for number of read plans requested
+    segment_read_plan(SegRefs, Indexes, []).
+
+-spec exec_read_plan(file:filename_all(), read_plan(), undefined | ra_flru:state(),
+                     TransformFun :: fun(),
+                     #{ra_index() => Command :: term()}) ->
+    {#{ra_index() => Command :: term()}, ra_flru:state()}.
+exec_read_plan(Dir, Plan, undefined, TransformFun, Acc0) ->
+    Open = ra_flru:new(1, fun({_, Seg}) -> ra_log_segment:close(Seg) end),
+    exec_read_plan(Dir, Plan, Open, TransformFun, Acc0);
+exec_read_plan(Dir, Plan, Open0, TransformFun, Acc0)
+  when is_list(Plan) ->
+    Fun = fun (I, T, B, Acc) ->
+                  E = TransformFun(I, T, binary_to_term(B)),
+                  Acc#{I => E}
+          end,
+    lists:foldl(
+      fun ({Idxs, BaseName}, {Acc1, Open1}) ->
+              {Seg, Open} = get_segment_ext(Dir, Open1, BaseName),
+              {_, Acc} = ra_log_segment:read_sparse(Seg, Idxs, Fun, Acc1),
+              {Acc, Open}
+      end, {Acc0, Open0}, Plan).
+
 -spec fetch_term(ra_index(), state()) -> {option(ra_index()), state()}.
 fetch_term(Idx, #?STATE{cfg = #cfg{} = Cfg} = State0) ->
     incr_counter(Cfg, ?C_RA_LOG_FETCH_TERM, 1),
     segment_term_query(Idx, State0).
 
 %% LOCAL
 
+segment_read_plan(_RegRefs, [], Acc) ->
+    lists:reverse(Acc);
+segment_read_plan([], _Indexes, Acc) ->
+    %% TODO: not all indexes were found
```

Comment:
A warning or error log might make sense here, wdyt?

Reply:
It may be easier for the caller to do a meaningful log message.